interfaces/cython/cantera/func1.pyx

```diff
@@ -83,7 +95,83 @@ cdef class Func1:
         return self.func.eval(t)
 
     def __reduce__(self):
-        raise NotImplementedError('Func1 object is not picklable')
+        msg = "'{}' objects are not picklable".format(type(self).__name__)
+        raise NotImplementedError(msg)
 
     def __copy__(self):
-        raise NotImplementedError('Func1 object is not copyable')
+        msg = "'{}' objects are not copyable".format(type(self).__name__)
+        raise NotImplementedError(msg)
+
+
+cdef class TabulatedFunction(Func1):
+    """
+    A `TabulatedFunction` object representing a tabulated function is defined by
+    sample points and corresponding function values. Inputs are specified either
+    by two iterable objects containing sample point location and function
+    values, or a single array that concatenates those inputs in two rows or
+    columns. Between sample points, values are evaluated based on the optional
+    keyword argument 'method'; options are 'linear' (linear interpolation,
+    default) or 'previous' (nearest previous value). Outside the sample
+    interval, the value at the closest end point is returned.
+
+    Examples for `TabulatedFunction` objects are::
+
+        >>> t1 = TabulatedFunction([[0, 2], [1, 1], [2, 0]])
+        >>> [t1(v) for v in [-0.5, 0, 0.5, 1.5, 2, 2.5]]
+        [2.0, 2.0, 1.5, 0.5, 0.0, 0.0]
+
+        >>> t2 = TabulatedFunction([[0, 2], [1, 1], [2, 0]], method='previous')
+        >>> [t2(v) for v in [-0.5, 0, 0.5, 1.5, 2, 2.5]]
+        [2.0, 2.0, 2.0, 1.0, 0.0, 0.0]
+
+        >>> t3 = TabulatedFunction([0, 1, 2], [2, 1, 0])
+        >>> [t3(v) for v in [-0.5, 0, 0.5, 1.5, 2, 2.5]]
+        [2.0, 2.0, 1.5, 0.5, 0.0, 0.0]
+
+        >>> t4 = TabulatedFunction(np.array([0, 1, 2]), (2, 1, 0))
+        >>> [t4(v) for v in [-0.5, 0, 0.5, 1.5, 2, 2.5]]
+        [2.0, 2.0, 1.5, 0.5, 0.0, 0.0]
```

Comment:
Can you provide prose explanations of what these are all doing, and what the difference between them are? 

Technically, the first and second cases here don't match the docstring which states that a single argument must be an array. Not sure there's a better way to phrase that though, just wanted to point it out in case you had any ideas.

Reply:
I guess I use *array* loosely here, as it can be a `numpy.array` (otherwise it's technically a `list`?). It also supports tuples ... also not sure, but I believe the examples are clearer now.